File: 1301-1400/1312-Minimum Insertion Steps to Make a String Palindrome/1312-Minimum Insertion Steps to Make a String Palindrome.cpp
class Solution {
public:
    int minInsertions(string s) {
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n));
        for (int j = 0; j < n; ++j) {
            dp[j][j] = 1;
            for (int i = j - 1; i >= 0; --i) {
                if (s[i] == s[j]) {
                    dp[i][j] = 2 + dp[i + 1][j - 1];
                }
                else {
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1]);
                }
            }
        }
        
        return n - dp[0][n - 1];
    }
};

File: 1301-1400/1312-Minimum Insertion Steps to Make a String Palindrome/1312-Minimum Insertion Steps to Make a String Palindrome.py
class Solution:
    def minInsertions(self, s: str) -> int:
        dp = [[0]*len(s) for _ in range(len(s))]
        for j in range(len(s)):
            dp[j][j] = 1
            for i in range(j - 1, -1, -1):
                if s[i] == s[j]:
                    dp[i][j] = 2 + dp[i + 1][j - 1]
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1])
        return len(s) - dp[0][-1]
